{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7930b5c5",
   "metadata": {},
   "source": [
    "# Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd9e53a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import gzip\n",
    "import tqdm.notebook as tqdm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436bd4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "DISABLE_PBAR = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3e004b",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQDIR = \"/Users/addisonhowe/mount_point2/sequencing\"\n",
    "DATDIRBASE = \"/Users/addisonhowe/mount_point/extracted_samples/raw_data_alignment\"\n",
    "\n",
    "sam_header_fname = \"header.sam\"\n",
    "\n",
    "outdir = f\"../out/ecoli_coverage/arrays\"\n",
    "imgdir = f\"../out/ecoli_coverage/images\"\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "os.makedirs(imgdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30dfcb1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need to fix this name since in the basecov file there is a comma.\n",
    "FIX_NAMES = {\n",
    "    \"NC_010943.1_Stenotrophomonas_maltophilia_K279a__strain_K279a\":\n",
    "        \"NC_010943.1_Stenotrophomonas_maltophilia_K279a,_strain_K279a\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12c78cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "RAWNAME_TO_SAMPLE = {}\n",
    "SAMPLE_TO_RAWNAME = {}\n",
    "with open(f\"{SEQDIR}/raw_data_name_mapping.tsv\", \"r\") as f:\n",
    "    csvreader = csv.reader(f, delimiter=\"\\t\")\n",
    "    for row in csvreader:  # process each row\n",
    "        RAWNAME_TO_SAMPLE[row[0]] = row[1]\n",
    "        SAMPLE_TO_RAWNAME[row[1]] = row[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "524609a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "REGION_MAPPING = {}  # Map name to gene to interval\n",
    "\n",
    "for d in os.listdir(\"../igv/search_results\"):\n",
    "    search_file = f\"../igv/search_results/{d}\"\n",
    "    with open(search_file, \"r\") as f:\n",
    "        for line in f:\n",
    "            # Skip comment lines (optional)\n",
    "            if line.startswith(\"#\") or not line.strip():\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            if len(parts) >= 5:\n",
    "                seqname = parts[0]\n",
    "                start = int(parts[3])\n",
    "                end = int(parts[4])\n",
    "                if seqname not in REGION_MAPPING:\n",
    "                    REGION_MAPPING[seqname] = {}\n",
    "                if d not in REGION_MAPPING[seqname]:\n",
    "                    REGION_MAPPING[seqname][d] = []\n",
    "                REGION_MAPPING[seqname][d].append((start, end))\n",
    "\n",
    "for k in REGION_MAPPING:\n",
    "    print(k, REGION_MAPPING[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe60377",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_list = [f for f in os.listdir(outdir) if f.endswith(\".npz\")]\n",
    "key = \"Lee_A8Q_1_Ecoli_contig_1_polypolish\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2539fb8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_regions(x, dx=1):\n",
    "    regions = []\n",
    "    r0 = x[0]\n",
    "    r1 = np.nan\n",
    "    idx_prev = x[0]\n",
    "    for i in range(1, len(x)):\n",
    "        idx = x[i]\n",
    "        # Check if the current value is an extension of the current region\n",
    "        if idx - idx_prev <= dx:\n",
    "            # Extend the region\n",
    "            r1 = idx\n",
    "        else:\n",
    "            # Reached end of regions. Store and reset.\n",
    "            regions.append([r0, r1])\n",
    "            r0 = idx\n",
    "            r1 = np.nan\n",
    "        idx_prev = idx\n",
    "    # Append final region\n",
    "    regions.append([r0, r1])\n",
    "    return np.array(regions)\n",
    "\n",
    "def get_spike_height(x, regions):\n",
    "    spike_heights = np.zeros(len(regions))\n",
    "    for i, region in enumerate(regions):\n",
    "        r0, r1 = region\n",
    "        if np.isnan(r1):\n",
    "            r1 = r0\n",
    "        heights = x[int(r0):int(r1 + 1)]\n",
    "        spike_heights[i] = np.max(heights)\n",
    "    return spike_heights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e7b705",
   "metadata": {},
   "outputs": [],
   "source": [
    "kappa = 5  # Track regions exceeding kappa * median reads\n",
    "dx = 100  # Padding for regions used in `find_regions`\n",
    "\n",
    "for array in tqdm.tqdm(array_list, total=len(array_list), disable=DISABLE_PBAR):\n",
    "    x = np.load(f\"{outdir}/{array}\", allow_pickle=True)[key]\n",
    "\n",
    "    med = np.median(x)  # store the median value\n",
    "    spike_locations = np.argwhere(x > kappa * med).flatten()\n",
    "    spike_regions = find_regions(spike_locations, dx=dx)\n",
    "    spike_heights = get_spike_height(x, spike_regions)\n",
    "    np.save(\n",
    "        f\"{outdir}/{array.replace(\".npz\", \"\")}_spike_regions.npy\", \n",
    "        spike_regions\n",
    "    )\n",
    "    np.save(\n",
    "        f\"{outdir}/{array.replace(\".npz\", \"\")}_spike_heights.npy\", \n",
    "        spike_heights\n",
    "    )\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(8,6))\n",
    "    # Plot base coverage\n",
    "    ax.plot(np.arange(1, len(x) + 1), x, '.')\n",
    "    \n",
    "    # Add annotations\n",
    "    annotations = REGION_MAPPING.get(key, {})\n",
    "    ylims = ax.get_ylim()\n",
    "    labelheight = ylims[0] + 0.1 * (ylims[1] - ylims[0])\n",
    "    labelincrement = 0.1\n",
    "    for gene in annotations:\n",
    "        prev_start, prev_end = -1, -1\n",
    "        regions = annotations[gene]\n",
    "        for region in regions:\n",
    "            start, end = region\n",
    "            if prev_start != start or prev_end != end:\n",
    "                ax.axvline(start, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                ax.axvline(end, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                tbox = ax.text(start, labelheight, gene)\n",
    "                labelheight += 0.025 * (ylims[1] - ylims[0])\n",
    "            prev_start, prev_end = start, end\n",
    "    \n",
    "    ax.set_xlabel(f\"position\")\n",
    "    ax.set_ylabel(\"read count\")\n",
    "    ax.set_title(array)\n",
    "\n",
    "    plt.savefig(f\"{imgdir}/{array.replace(\".npz\", \"\")}.png\")\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b30c018c",
   "metadata": {},
   "outputs": [],
   "source": [
    "bin_frac = 0.001\n",
    "for array in tqdm.tqdm(array_list, total=len(array_list), disable=DISABLE_PBAR):\n",
    "    x = np.load(f\"{outdir}/{array}\", allow_pickle=True)[key]\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(8,6))\n",
    "    n = len(x)\n",
    "    binsize = int(np.ceil(bin_frac * n))\n",
    "    nbins = n // binsize + (n % binsize != 0)\n",
    "    remainder = n % binsize\n",
    "    if remainder > 0:\n",
    "        x = np.concatenate([x, np.zeros(binsize - remainder)])\n",
    "    xmaxed = x.reshape([-1, binsize]).max(axis=1)\n",
    "    # Plot base coverage\n",
    "    ax.plot(np.arange(1, len(xmaxed) + 1), xmaxed, '.')\n",
    "    \n",
    "    # Add annotations\n",
    "    annotations = REGION_MAPPING.get(key, {})\n",
    "    ylims = ax.get_ylim()\n",
    "    labelheight = ylims[0] + 0.1 * (ylims[1] - ylims[0])\n",
    "    labelincrement = 0.1\n",
    "    for gene in annotations:\n",
    "        prev_start, prev_end = -1, -1\n",
    "        regions = annotations[gene]\n",
    "        for region in regions:\n",
    "            start, end = region\n",
    "            start = 1 + start // binsize\n",
    "            end = 1 + end // binsize\n",
    "            if prev_start != start or prev_end != end:\n",
    "                ax.axvline(start, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                ax.axvline(end, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                tbox = ax.text(start, labelheight, gene)\n",
    "                labelheight += 0.025 * (ylims[1] - ylims[0])\n",
    "            prev_start, prev_end = start, end\n",
    "    \n",
    "    ax.set_xlabel(f\"bin (size {binsize})\")\n",
    "    ax.set_ylabel(\"read count\")\n",
    "    ax.set_title(array)\n",
    "\n",
    "    plt.savefig(f\"{imgdir}/{array.replace(\".npz\", \"\")}_binned.png\")\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bf39d8e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
