{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7930b5c5",
   "metadata": {},
   "source": [
    "# Spike-in coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd9e53a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import csv\n",
    "import tqdm.notebook as tqdm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436bd4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "SOIL_KEY = \"Soil11\"\n",
    "DISABLE_PBAR = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3e004b",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQDIR = \"/Users/addisonhowe/mnts/raw_data/sequencing\"\n",
    "DATDIR = f\"{SEQDIR}/coverage/arrays\"  # location of .npz soil coverage arrays\n",
    "sam_header_fname = \"header.sam\"\n",
    "\n",
    "OUTDIR = \"../out/coverage\"\n",
    "os.makedirs(OUTDIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30dfcb1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need to fix this name since in the basecov file there is a comma.\n",
    "FIX_NAMES = {\n",
    "    \"NC_010943.1_Stenotrophomonas_maltophilia_K279a__strain_K279a\":\n",
    "        \"NC_010943.1_Stenotrophomonas_maltophilia_K279a,_strain_K279a\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12c78cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "RAWNAME_TO_SAMPLE = {}\n",
    "SAMPLE_TO_RAWNAME = {}\n",
    "with open(f\"{SEQDIR}/raw_data_name_mapping.tsv\", \"r\") as f:\n",
    "    csvreader = csv.reader(f, delimiter=\"\\t\")\n",
    "    for row in csvreader:  # process each row\n",
    "        RAWNAME_TO_SAMPLE[row[0]] = row[1]\n",
    "        SAMPLE_TO_RAWNAME[row[1]] = row[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "524609a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "REGION_MAPPING = {}  # Map name to gene to interval\n",
    "\n",
    "for d in os.listdir(\"../igv/out/search_results\"):\n",
    "    search_file = f\"../igv/out/search_results/{d}\"\n",
    "    with open(search_file, \"r\") as f:\n",
    "        for line in f:\n",
    "            # Skip comment lines (optional)\n",
    "            if line.startswith(\"#\") or not line.strip():\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            if len(parts) >= 5:\n",
    "                seqname = parts[0]\n",
    "                start = int(parts[3])\n",
    "                end = int(parts[4])\n",
    "                if seqname not in REGION_MAPPING:\n",
    "                    REGION_MAPPING[seqname] = {}\n",
    "                if d not in REGION_MAPPING[seqname]:\n",
    "                    REGION_MAPPING[seqname][d] = []\n",
    "                REGION_MAPPING[seqname][d].append((start, end))\n",
    "\n",
    "for k in REGION_MAPPING:\n",
    "    print(k, REGION_MAPPING[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "979b2311",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_SOIL_SAMPLES = [f for f in os.listdir(DATDIR) if f.endswith(\".npz\")]\n",
    "\n",
    "def order_samples(sample_list, searchkey):\n",
    "    if searchkey:\n",
    "        subset = [s for s in sample_list if searchkey + \"_\" in s]\n",
    "    else:\n",
    "        subset = sample_list\n",
    "\n",
    "    t0_set = [s for s in subset if \"T0\" in s]\n",
    "    t9_set = [s for s in subset if \"T9\" in s]\n",
    "    no_nitrate_t0_set = [s for s in t0_set if \"No_Nitrate\" in s]\n",
    "    no_nitrate_t9_set = [s for s in t9_set if \"No_Nitrate\" in s]\n",
    "    t0_set = [s for s in t0_set if \"No_Nitrate\" not in s]\n",
    "    t9_set = [s for s in t9_set if \"No_Nitrate\" not in s]\n",
    "    chl_set = [s for s in t9_set if \"CHL\" in s]\n",
    "    none_set = [s for s in t9_set if \"None\" in s]\n",
    "    \n",
    "    def sortfunc(s):\n",
    "        sp = s.split(\"_\")\n",
    "        if sp[5][0] == \"N\":\n",
    "            v = float(sp[6])\n",
    "        else:\n",
    "            v = float(sp[5])\n",
    "        return v\n",
    "\n",
    "    chl_set = sorted(chl_set, key=sortfunc)\n",
    "    none_set = sorted(none_set, key=sortfunc)\n",
    "    no_nitrate_t0_set = sorted(no_nitrate_t0_set, key=sortfunc)\n",
    "    no_nitrate_t9_set = sorted(no_nitrate_t9_set, key=sortfunc)\n",
    "    ord = t0_set + chl_set + none_set + no_nitrate_t0_set + no_nitrate_t9_set\n",
    "    assert set(ord) == set(subset) and len(ord) == len(subset), \\\n",
    "        f\"Messed up! Expected {subset}. Got {ord}\"\n",
    "    return ord\n",
    "\n",
    "\n",
    "SAMPLES_BY_SOIL = {}\n",
    "for i in [11, 12, 14, 15, 16, 17, 3, 5, 6, 9]:\n",
    "    k = f\"Soil{i}\"\n",
    "    SAMPLES_BY_SOIL[k.lower()] =  order_samples(ALL_SOIL_SAMPLES, k)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe60377",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list = SAMPLES_BY_SOIL[SOIL_KEY.lower()]\n",
    "\n",
    "for s in sample_list:\n",
    "    print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2539fb8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_spiked_regions(x, dx=1):\n",
    "    \"\"\"Identify spiked regions.\"\"\"\n",
    "    if len(x) == 0:\n",
    "        return []\n",
    "    regions = []\n",
    "    r0 = x[0]\n",
    "    r1 = x[0] + 1\n",
    "    idx_prev = x[0]\n",
    "    for i in range(1, len(x)):\n",
    "        idx = x[i]\n",
    "        # Check if the current value is an extension of the current region\n",
    "        if idx - idx_prev <= dx:\n",
    "            # Extend the region\n",
    "            r1 = idx + 1\n",
    "        else:\n",
    "            # Reached end of regions. Store and reset.\n",
    "            regions.append([r0, r1])\n",
    "            r0 = idx\n",
    "            r1 = idx + 1\n",
    "        idx_prev = idx\n",
    "    # Append final region\n",
    "    regions.append([r0, r1])\n",
    "    return np.array(regions)\n",
    "\n",
    "def get_spike_height(x, regions):\n",
    "    spike_heights = np.zeros(len(regions))\n",
    "    for i, region in enumerate(regions):\n",
    "        r0, r1 = region\n",
    "        if np.isnan(r1):\n",
    "            r1 = r0\n",
    "        heights = x[int(r0):int(r1 + 1)]\n",
    "        spike_heights[i] = np.max(heights)\n",
    "    return spike_heights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3dae88",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################################################\n",
    "##  Tests for find_spiked_regions\n",
    "\n",
    "x = np.concatenate([\n",
    "    np.zeros(10),\n",
    "    np.zeros(10),\n",
    "    np.arange(6),\n",
    "    np.flip(np.arange(5)),\n",
    "    np.zeros(10),\n",
    "    np.zeros(10),\n",
    "    np.zeros(10),\n",
    "    np.arange(4)\n",
    "], dtype=float)\n",
    "\n",
    "kappa = 5\n",
    "dx = 1\n",
    "print(len(x))\n",
    "med = np.median(x)\n",
    "print(f\"kappa={kappa}, med={med}\")\n",
    "spike_locations = np.argwhere(x > kappa * med).flatten()\n",
    "print(\"Spike locations:\", spike_locations)\n",
    "regs = find_spiked_regions(spike_locations, dx=dx)\n",
    "print(\"spike regions:\\n\", regs)\n",
    "heights = get_spike_height(x, regs)\n",
    "print(\"spike heights:\\n\", heights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e7b705",
   "metadata": {},
   "outputs": [],
   "source": [
    "dx = 200  # Padding for regions used in `find_spiked_regions`\n",
    "bin_target = 500\n",
    "logscale=lambda x: np.where(x == 0, np.nan, np.log2(x, where=(x != 0)))\n",
    "\n",
    "spike_marker_size = None\n",
    "\n",
    "for sampidx, sample in tqdm.tqdm(\n",
    "        enumerate(sample_list),\n",
    "        total=len(sample_list), \n",
    "        desc=\"Sample\",\n",
    "        leave=False,\n",
    "        disable=DISABLE_PBAR,\n",
    "):\n",
    "    arr = np.load(f\"{DATDIR}/{sample}\", allow_pickle=True)\n",
    "    scaffold_keys = list(arr.keys())\n",
    "    for scaffold_key in tqdm.tqdm(\n",
    "            scaffold_keys, \n",
    "            total=len(scaffold_keys), \n",
    "            desc=\"scaffold\", \n",
    "            leave=False,\n",
    "            disable=DISABLE_PBAR,\n",
    "    ):\n",
    "        x = arr[scaffold_key]\n",
    "        outdir = f\"{OUTDIR}/by_scaffold/{scaffold_key}/arrays\"\n",
    "        imgdir = f\"{OUTDIR}/by_scaffold/{scaffold_key}/images\"\n",
    "        os.makedirs(outdir, exist_ok=True)\n",
    "        os.makedirs(imgdir, exist_ok=True)\n",
    "\n",
    "        #####################################################################\n",
    "        ##  Compute and save spike information, etc.\n",
    "\n",
    "        bin_frac = bin_target / len(x)\n",
    "        xpos = x[x > 0]\n",
    "        med = np.median(xpos)  # store the median value\n",
    "        meanlog = np.mean(np.log2(xpos))  # store the mean log2 of positives\n",
    "        stdlog = np.std(np.log2(xpos))  # store the std log2 of positives\n",
    "        p25 = np.percentile(xpos, 25)  # store the 25th percentile\n",
    "        p75 = np.percentile(xpos, 75)  # store the 75th percentile\n",
    "        iqr = p75 - p25  # compute the IQR\n",
    "        # threshold = med + 1.5 * iqr\n",
    "        threshold = 2**(meanlog + 2.5 * stdlog)\n",
    "        spike_locations = np.argwhere(x > threshold).flatten()\n",
    "        spike_regions = find_spiked_regions(spike_locations, dx=dx)\n",
    "        spike_heights = get_spike_height(x, spike_regions)\n",
    "        np.save(\n",
    "            f\"{outdir}/{sample.replace(\".npz\", \"\")}_spike_regions.npy\", \n",
    "            spike_regions\n",
    "        )\n",
    "        np.save(\n",
    "            f\"{outdir}/{sample.replace(\".npz\", \"\")}_spike_heights.npy\", \n",
    "            spike_heights\n",
    "        )\n",
    "\n",
    "        # Compute bins\n",
    "        n = len(x)\n",
    "        binsize = int(np.ceil(bin_frac * n))\n",
    "        nbins = n // binsize + (n % binsize != 0)\n",
    "        remainder = n % binsize\n",
    "        if remainder > 0:\n",
    "            x = np.concatenate([x, np.zeros(binsize - remainder)])\n",
    "        xbinned = x.reshape([-1, binsize]).max(axis=1)\n",
    "\n",
    "        #####################################################################\n",
    "        ##  Plot data\n",
    "\n",
    "        fig, ax = plt.subplots(1, 1, figsize=(10,6))\n",
    "\n",
    "        # Plot base coverage\n",
    "        ts = 1 + np.arange(0, len(xbinned))\n",
    "        ax.plot(\n",
    "            ts, logscale(xbinned), \n",
    "            linestyle='None',\n",
    "            marker=\".\", \n",
    "            markersize=spike_marker_size,\n",
    "        )\n",
    "\n",
    "        ax.set_xlim(-ts.max() * 0.01, ts.max() * 1.05)\n",
    "        # ax.set_ylim(-0.01, ax.get_ylim()[1])\n",
    "\n",
    "        # Mark median and threshold\n",
    "        xlims = ax.get_xlim()\n",
    "        # ax.hlines(\n",
    "        #     logscale(med), *xlims, \n",
    "        #     linestyle=':', linewidth=1, color=\"r\",\n",
    "        #     label=f\"med\",\n",
    "        # )\n",
    "        # ax.hlines(\n",
    "        #     logscale(p75), *xlims, \n",
    "        #     linestyle=\"--\", linewidth=1, color=\"r\",\n",
    "        #     label=\"75th pctl\",\n",
    "        # )\n",
    "        ax.hlines(\n",
    "            logscale(threshold), *xlims, \n",
    "            linestyle=\"-\", linewidth=1, color=\"r\",\n",
    "            label=\"threshold\",\n",
    "        )\n",
    "        ax.set_xlim(*xlims)\n",
    "\n",
    "        # Mark spike regions\n",
    "        ylims = ax.get_ylim()\n",
    "        for reg, height in zip(spike_regions, spike_heights):\n",
    "            if height >= threshold:\n",
    "                ax.axvspan(\n",
    "                    reg[0], reg[1],*ylims, \n",
    "                    color='orange', alpha=0.5,\n",
    "                )\n",
    "        ax.set_ylim(*ylims)\n",
    "        \n",
    "        # Add annotations\n",
    "        annotations = REGION_MAPPING.get(scaffold_key, {})\n",
    "        ylims = ax.get_ylim()\n",
    "        labelheight = ylims[1]\n",
    "        for gene in annotations:\n",
    "            prev_start, prev_end = -1, -1\n",
    "            regions = annotations[gene]\n",
    "            for region in regions:\n",
    "                start, end = region\n",
    "                if prev_start != start or prev_end != end:\n",
    "                    ax.axvline(start, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                    ax.axvline(end, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                    tbox = ax.text(\n",
    "                        start, labelheight, gene.split(\"_\")[1],\n",
    "                        verticalalignment=\"top\",\n",
    "                    )\n",
    "                    labelheight -= 0.025 * (ylims[1] - ylims[0])\n",
    "                prev_start, prev_end = start, end\n",
    "        \n",
    "        ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "        ax.set_xlabel(f\"position\")\n",
    "        ax.set_ylabel(\"log2 read count\")\n",
    "        title = scaffold_key\n",
    "        subtitle = sample.removeprefix(\"coverage_arrays_\").removesuffix(\".npz\")\n",
    "        ax.set_title(title + \"\\n\" + subtitle)\n",
    "\n",
    "        plt.savefig(f\"{imgdir}/{sample.replace(\".npz\", \"\")}.png\", \n",
    "                    bbox_inches=\"tight\")\n",
    "        plt.close()\n",
    "\n",
    "        #####################################################################\n",
    "        ##  Plot binned data\n",
    "        \n",
    "        fig, ax = plt.subplots(1, 1, figsize=(10,6))\n",
    "        \n",
    "        # Plot base coverage\n",
    "        ts = 1 + np.arange(0, len(xbinned))\n",
    "        ax.plot(\n",
    "            ts, logscale(xbinned), \n",
    "            linestyle='None',\n",
    "            marker=\".\", \n",
    "            markersize=spike_marker_size,\n",
    "        )\n",
    "\n",
    "        ax.set_xlim(-ts.max() * 0.01, ts.max() * 1.05)\n",
    "        # ax.set_ylim(-0.01, ax.get_ylim()[1])\n",
    "\n",
    "        # Mark median and threshold\n",
    "        xlims = ax.get_xlim()\n",
    "        # ax.hlines(\n",
    "        #     logscale(med), *xlims, \n",
    "        #     linestyle=':', linewidth=1, color=\"r\",\n",
    "        #     label=\"med\",\n",
    "        # )\n",
    "        # ax.hlines(\n",
    "        #     logscale(p75), *xlims, \n",
    "        #     linestyle=\"--\", linewidth=1, color=\"r\",\n",
    "        #     label=\"75th pctl\",\n",
    "        # )\n",
    "        ax.hlines(\n",
    "            logscale(threshold), *xlims, \n",
    "            linestyle=\"-\", linewidth=1, color=\"r\",\n",
    "            label=\"threshold\",\n",
    "        )\n",
    "        ax.set_xlim(*xlims)\n",
    "\n",
    "        # Mark spike regions\n",
    "        ylims = ax.get_ylim()\n",
    "        for reg, height in zip(spike_regions, spike_heights):\n",
    "            if height >= threshold:\n",
    "                r0 = reg[0] // binsize\n",
    "                r1 = reg[1] // binsize\n",
    "                ax.axvspan(\n",
    "                    r0, r1, *ylims, \n",
    "                    color='orange', alpha=0.5,\n",
    "                )\n",
    "        ax.set_ylim(*ylims)\n",
    "        \n",
    "        # Add annotations\n",
    "        annotations = REGION_MAPPING.get(scaffold_key, {})\n",
    "        ylims = ax.get_ylim()\n",
    "        labelheight = ylims[1]\n",
    "        for gene in annotations:\n",
    "            prev_start, prev_end = -1, -1\n",
    "            regions = annotations[gene]\n",
    "            for region in regions:\n",
    "                start, end = region\n",
    "                start = 1 + start // binsize\n",
    "                end = 1 + end // binsize\n",
    "                if prev_start != start or prev_end != end:\n",
    "                    ax.axvline(start, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                    ax.axvline(end, *ylims, linestyle=\":\", color=\"k\", zorder=1)\n",
    "                    tbox = ax.text(\n",
    "                        start, labelheight, gene.split(\"_\")[1],\n",
    "                        verticalalignment=\"top\",\n",
    "                    )\n",
    "                    labelheight -= 0.025 * (ylims[1] - ylims[0])\n",
    "                prev_start, prev_end = start, end\n",
    "        \n",
    "        ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "        ax.set_xlabel(f\"bin (size {bin_frac*100:.2g}%={binsize}bp)\")\n",
    "        ax.set_ylabel(\"log2 read count\")\n",
    "        title = scaffold_key\n",
    "        subtitle = sample.removeprefix(\"coverage_arrays_\").removesuffix(\".npz\")\n",
    "        ax.set_title(title + \"\\n\" + subtitle)\n",
    "\n",
    "        plt.savefig(f\"{imgdir}/{sample.replace(\".npz\", \"\")}_binned.png\", \n",
    "                    bbox_inches=\"tight\")\n",
    "        plt.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77352e33",
   "metadata": {},
   "source": [
    "#### Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77bec14a",
   "metadata": {},
   "outputs": [],
   "source": [
    "SCAFFOLD_KEYS = [\n",
    "    \"Ralstonia_solanacearum_strain_KACC_10722\",\n",
    "    \"CP013959.1_Staphylococcus_aureus_strain_V605\",\n",
    "    \"NC_003902.1_Xanthomonas_campestris_pv._campestris_str._ATCC_33913\",\n",
    "    \"NC_003295.1_Ralstonia_solanacearum_GMI1000\",\n",
    "    \"NC_012660.1_Pseudomonas_fluorescens_SBW25\",\n",
    "    \"NC_002570.2_Bacillus_halodurans_C-125\",\n",
    "    \"NC_002976.3_Staphylococcus_epidermidis_RP62A\",\n",
    "    \"NC_007492.2_Pseudomonas_fluorescens_Pf0-1\",\n",
    "    \"NC_008782.1_Acidovorax_sp._JS42\",\n",
    "    \"NC_010002.1_Delftia_acidovorans_SPH-1\",\n",
    "    \"NC_011071.1_Stenotrophomonas_maltophilia_R551-3\",\n",
    "    \"NC_008702.1_Azoarcus_sp._BH72\",\n",
    "    \"NC_010943.1_Stenotrophomonas_maltophilia_K279a,_strain_K279a\",\n",
    "    \"NC_013446.2_Comamonas_testosteroni_CNB-2\",\n",
    "    \"NC_014323.1_Herbaspirillum_seropedicae_SmR1\",\n",
    "    \"NC_015563.1_Delftia_sp._Cs1-4\",\n",
    "    \"NC_016830.1_Pseudomonas_fluorescens_F113\",\n",
    "    \"NZ_AJVM01000001.1_Rhizobium_sp._AP16\",\n",
    "    \"NC_018708.1_Acidovorax_sp._KKS102\",\n",
    "    \"NC_020561.1_Sphingomonas_sp._MM-1\",\n",
    "    \"NZ_KB906253.1_Curvibacter_lanceolatus_ATCC_14669\",\n",
    "    \"NZ_KB976985.1_Acinetobacter_calcoaceticus_ANC_3811\",\n",
    "    \"NZ_KI421535.1_Burkholderia_cepacia_ATCC_25416\",\n",
    "    \"NZ_HG916852.1_Rhizobium_sp._LPU83\",\n",
    "    \"NZ_CP007638.1_Pseudomonas_sp._WCS374\",\n",
    "    \"NZ_CP008788.1_Klebsiella_oxytoca_KONIH1\",\n",
    "    \"NZ_CP010536.1_Cupriavidus_basilensis_strain_4G11\",\n",
    "    \"NZ_JXYQ01000001.1_Acidovorax_temperans_strain_KY4\",\n",
    "    \"NZ_CP010516.1_Cupriavidus_gilardii_CR3\",\n",
    "    \"NZ_CP010350.1_Acinetobacter_johnsonii_XBB1\",\n",
    "    \"NZ_CP014540.1_Acinetobacter_baumannii_strain_XH857\",\n",
    "    \"NC_016845.1_Klebsiella_pneumoniae_subsp._pneumoniae_HS11286\",\n",
    "    \"NC_007606.1_Shigella_dysenteriae_Sd197\",\n",
    "    \"NC_004337.2_Shigella_flexneri_2a_str._301\",\n",
    "    \"Lee_A8Q_1_Ecoli_contig_1_polypolish\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93bab00",
   "metadata": {},
   "outputs": [],
   "source": [
    "dx = 200  # Padding for regions used in `find_spiked_regions`\n",
    "bin_target = 500\n",
    "\n",
    "CMAP = \"plasma\"\n",
    "spike_color = \"cyan\"\n",
    "spike_marker_size = 3\n",
    "\n",
    "imgdir = f\"{OUTDIR}/images/{SOIL_KEY.lower()}\"\n",
    "os.makedirs(imgdir, exist_ok=True)\n",
    "\n",
    "progbar = tqdm.tqdm(\n",
    "    SCAFFOLD_KEYS, total=len(SCAFFOLD_KEYS), \n",
    "    desc=\"scaffold\", \n",
    "    leave=False,\n",
    "    disable=DISABLE_PBAR,\n",
    ")\n",
    "for scaffold_key in progbar:\n",
    "    progbar.set_description(scaffold_key)\n",
    "    \n",
    "    all_x = []\n",
    "    all_medians = []\n",
    "    all_meanlogs = []\n",
    "    all_stdlogs = []\n",
    "    all_p75s = []\n",
    "    all_p25s = []\n",
    "    all_iqrs = []\n",
    "    all_thresholds = []\n",
    "    all_spike_locations = []\n",
    "    all_spike_regions = []\n",
    "    all_spike_heights = []\n",
    "    all_xbinned = []\n",
    "\n",
    "    for sample in tqdm.tqdm(\n",
    "        sample_list, \n",
    "        total=len(sample_list), \n",
    "        disable=DISABLE_PBAR,\n",
    "        desc=\"Sample\",\n",
    "        leave=False,\n",
    "    ):\n",
    "        arr = np.load(f\"{DATDIR}/{sample}\", allow_pickle=True)\n",
    "        x = arr[scaffold_key]\n",
    "\n",
    "        #####################################################################\n",
    "        ##  Compute and store spike information, etc.\n",
    "        xpos = x[x > 0]\n",
    "        med = np.median(xpos)  # store the median value\n",
    "        p75 = np.percentile(xpos, 75)  # store the 75th percentile\n",
    "        p25 = np.percentile(xpos, 25)  # store the 25th percentile\n",
    "        meanlog = np.mean(np.log2(xpos))  # store the mean log2 of positives\n",
    "        stdlog = np.std(np.log2(xpos))  # store the std log2 of positives\n",
    "        iqr = p75 - p25\n",
    "        # threshold = med + 1.5 * iqr\n",
    "        threshold = 2**(meanlog + 2.5 * stdlog)\n",
    "        spike_locations = np.argwhere(x > threshold).flatten()\n",
    "        spike_regions = find_spiked_regions(spike_locations, dx=dx)\n",
    "        spike_heights = get_spike_height(x, spike_regions)\n",
    "\n",
    "        all_x.append(x)\n",
    "        all_medians.append(med)\n",
    "        all_p75s.append(p75)\n",
    "        all_p25s.append(p25)\n",
    "        all_meanlogs.append(meanlog)\n",
    "        all_stdlogs.append(stdlog)\n",
    "        all_iqrs.append(iqr)\n",
    "        all_thresholds.append(threshold)\n",
    "        all_spike_locations.append(spike_locations)\n",
    "        all_spike_regions.append(spike_regions)\n",
    "        all_spike_heights.append(spike_heights)\n",
    "        \n",
    "        n = len(x)\n",
    "        bin_frac = bin_target / n\n",
    "        binsize = int(np.ceil(bin_frac * n))\n",
    "        nbins = n // binsize + (n % binsize != 0)\n",
    "        remainder = n % binsize\n",
    "        if remainder > 0:\n",
    "            x = np.concatenate([x, np.zeros(binsize - remainder)])\n",
    "        xbinned = x.reshape([-1, binsize]).max(axis=1)\n",
    "\n",
    "        all_xbinned.append(xbinned)\n",
    "\n",
    "    all_x = np.array(all_x)\n",
    "    all_medians = np.array(all_medians)\n",
    "    all_thresholds = np.array(all_thresholds)\n",
    "    all_p75s = np.array(all_p75s)\n",
    "    all_xbinned = np.array(all_xbinned)\n",
    "\n",
    "    #####################################################################\n",
    "    ##  Plot\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(13,7))\n",
    "    \n",
    "    sc = ax.imshow(\n",
    "        all_xbinned, \n",
    "        aspect='auto', \n",
    "        norm=\"log\",\n",
    "        cmap=CMAP, \n",
    "    )\n",
    "\n",
    "    divider = make_axes_locatable(ax)\n",
    "    cax = divider.append_axes('right', size='5%', pad=0.1)\n",
    "    cbar = fig.colorbar(sc, cax=cax)\n",
    "    cbar.ax.set_title(\"reads\", size=10)\n",
    "\n",
    "    ticks = np.arange(len(sample_list))\n",
    "    ticklabels = [str(i+1) for i in ticks]\n",
    "    annot_space = 1/3\n",
    "    ax.set_ylim(1/(1-annot_space) * ax.get_ylim()[0], ax.get_ylim()[1])\n",
    "    ax.set_yticks(ticks, labels=ticklabels)\n",
    "\n",
    "    # Add annotations\n",
    "    annotations = REGION_MAPPING.get(scaffold_key, {})\n",
    "    gene_order = sorted(\n",
    "        list(annotations.keys()), key=lambda g: annotations[g][0][0], \n",
    "    )\n",
    "    ylims = ax.get_ylim()\n",
    "    labelheight0 = ylims[1] - (1 - annot_space) * (ylims[1] - ylims[0]) + 0.1\n",
    "    labelheight = labelheight0\n",
    "    for gene in gene_order:\n",
    "        prev_start, prev_end = -1, -1\n",
    "        regions = annotations[gene]\n",
    "        for region in regions:\n",
    "            start, end = region\n",
    "            start = 1 + start // binsize\n",
    "            end = 1 + end // binsize\n",
    "            if prev_start != start or prev_end != end:\n",
    "                ax.vlines(\n",
    "                    start, ylims[0], (1 - annot_space) * ylims[0], \n",
    "                    linestyle=\"--\", color=\"k\", alpha=1, zorder=1,\n",
    "                )\n",
    "                tbox = ax.text(\n",
    "                    start, labelheight, gene.split(\"_\")[1],\n",
    "                    verticalalignment=\"top\", color='r', zorder=2,\n",
    "                )\n",
    "                labelheight -= 0.025 * (ylims[1] - ylims[0])\n",
    "                if labelheight - 0.025 * (ylims[1] - ylims[0]) >= ylims[0]:\n",
    "                    labelheight = labelheight0\n",
    "            prev_start, prev_end = start, end\n",
    "    \n",
    "    # Add spike markers\n",
    "    for sampidx, spike_regions in enumerate(all_spike_regions):\n",
    "        spike_heights = all_spike_heights[sampidx]\n",
    "        threshold = all_thresholds[sampidx]\n",
    "        for spike_region, spike_height in zip(spike_regions, spike_heights):\n",
    "            if spike_height < threshold:\n",
    "                continue\n",
    "            pos = np.mean(spike_region)\n",
    "            binpos = pos // binsize\n",
    "            ax.plot(\n",
    "                binpos, sampidx, \n",
    "                marker='.',\n",
    "                color=spike_color,\n",
    "                markersize=spike_marker_size,\n",
    "            )\n",
    "\n",
    "    soilnames = [\n",
    "        s.removeprefix(\"coverage_arrays_\").removesuffix(\".npz\").replace(\"_\", \" \") \n",
    "        for s in sample_list\n",
    "    ]\n",
    "    ax.set_yticklabels(soilnames, fontsize=8, rotation=45)\n",
    "    ax.set_xlabel(f\"bin (size {bin_frac*100:.2g}%={binsize}bp)\")\n",
    "    ax.set_ylabel(\"sample\")\n",
    "    title = scaffold_key\n",
    "    subtitle = SOIL_KEY\n",
    "    ax.set_title(title + \"\\n\" + subtitle)\n",
    "\n",
    "    plt.savefig(f\"{imgdir}/basecov_raw_{scaffold_key}.png\")\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc4ab91",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
